### PR TITLE
Feature/block priofees

### PR DESCRIPTION
**Pull request by grooviegermanikus**
### Local testing

use this logging settings:
> RUST_LOG=info,lite_rpc=debug,solana_lite_rpc_block_priofees=trace

tested against mainnet: `http://mango.rpcpool.com/.....`

Request:
```
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "getLatestBlockPrioFees",
  "params": []
}

```

Response:
```
{
  "jsonrpc": "2.0",
  "result": {
    "context": {
      "slot": 243836394
    },
    "value": {
      "by_tx": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        1,
        100,
        100,
        610,
        3151,
        4500,
        5075,
        9868,
        19736,
        30040,
        30124,
        50000,
        100000,
        8486574
      ],
      "by_tx_percentiles": [
        0,
        0.05,
        0.1,
        0.15,
        0.2,
        0.25,
        0.3,
        0.35,
        0.4,
        0.45,
        0.5,
        0.55,
        0.6,
        0.65,
        0.7,
        0.75,
        0.8,
        0.85,
        0.9,
        0.95,
        1
      ],
      "by_cu": [
        0,
        0,
        0,
        0,
        0,
        0,
        1,
        100,
        677,
        1000,
        2600,
        3600,
        5000,
        8333,
        10003,
        25000,
        30107,
        30120,
        50000,
        194939,
        8486574
      ],
      "by_cu_percentiles": [
        0,
        0.05,
        0.1,
        0.15,
        0.2,
        0.25,
        0.3,
        0.35,
        0.4,
        0.45,
        0.5,
        0.55,
        0.6,
        0.65,
        0.7,
        0.75,
        0.8,
        0.85,
        0.9,
        0.95,
        1
      ],
      "tx_count": {
        "total": 1501,
        "nonvote": 415
      },
      "cu_consumed": {
        "total": 42415372,
        "nonvote": 40134772
      }
    }
  },
  "id": 1
}
```

Websocket:
```
 Received Text: {"jsonrpc":"2.0","method":"blockPrioritizationFeesNotification","params":{"subscription":3026874038738617,"result":{"context":{"slot":243836441},"value":{"by_tx":[0,0,0,0,0,4,100,1000,1000,1000,3800,5000,9180,12500,20000,30130,40000,40000,50198,59077,179765586],"by_tx_percentiles":[0.0,0.05,0.1,0.15,0.2,0.25,0.3,0.35,0.4,0.45,0.5,0.55,0.6,0.65,0.7,0.75,0.8,0.85,0.9,0.95,1.0],"by_cu":[0,0,0,0,2,100,800,1000,1000,1000,1500,3700,5000,5990,10003,10003,24848,30143,40000,110000,179765586],"by_cu_percentiles":[0.0,0.05,0.1,0.15,0.2,0.25,0.3,0.35,0.4,0.45,0.5,0.55,0.6,0.65,0.7,0.75,0.8,0.85,0.9,0.95,1.0],"tx_count":{"total":1450,"nonvote":382},"cu_consumed":{"total":37565771,"nonvote":35322971}}}}}
```

### CODE REVIEW

File: block_priofees/src/block_priofees.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+use crate::stats_calculation::calculate_supp_stats;
+
+// note: ATM only the latest slot (highest key) is used
+const SLOTS_TO_RETAIN: u64 = 5000;
+
+/// put everything required to serve sync data calls here
+#[derive(Clone)]
+pub struct PrioFeeStore {
+    // store priofees stats for recently confirmed blocks up to CLEANUP_SLOTS_AFTER
+    recent: Arc<DashMap<Slot, PrioFeesStats>>,
```
Comment (godmodegalactus) [inline block_priofees/src/block_priofees.rs:19]: Use BTreeMap so that the max is always the last.

File: block_priofees/src/block_priofees.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                Ok(block) => {
+                    // first do some cleanup
+                    recent_data.retain(|slot, _| *slot > slot - SLOTS_TO_RETAIN);
+
+                    if !block.commitment_config.is_confirmed() {
+                        continue;
+                    }
+                    let confirmed_slot = block.slot;
+
+                    let block_prio_fees = block
```
Comment (godmodegalactus) [inline block_priofees/src/block_priofees.rs:63]: You have to ignore vote transactions.

File: block_priofees/src/stats_calculation.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+) -> PrioFeesStats {
+    let mut prio_fees_in_block = if prio_fees_in_block.is_empty() {
+        // TODO is that smart?
+        vec![(0, 0)]
+    } else {
+        prio_fees_in_block.clone()
+    };
+    prio_fees_in_block.sort_by(|a, b| a.0.cmp(&b.0));
+
+    let median_index = prio_fees_in_block.len() / 2;
```
Comment (godmodegalactus) [inline block_priofees/src/stats_calculation.rs:17]: The priority fees should be done by CU and transaction count.

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

File: block_priofees/src/block_priofees.rs (hunk 2 above)
Comment (grooviegermanikus) [reply block_priofees/src/block_priofees.rs]: okey .. but how?

Comment (grooviegermanikus) [reply block_priofees/src/block_priofees.rs]: look here: https://github.com/blockworks-foundation/lite-rpc/pull/274/commits/2a4e86177d6cd2fb04928ee9959e42cadbef05a9?

**Comment by mschneider** [comment]
> tested against mainnet: `https://api.testnet.rpcpool.com/`

the url is a testnet cluster, the comment says mainnet, please clarify

**Comment by mschneider** [comment]
when i just tried to connect i wasn't able to open a websocket connection

```
mxa:mango-router max$ yarn wscat --connect wss://api.mngo.cloud/lite-rpc-staging/v1
yarn run v1.22.15
$ /Users/max/Code/mango-router/node_modules/.bin/wscat --connect wss://api.mngo.cloud/lite-rpc-staging/v1
error: Unexpected server response: 200
error Command failed with exit code 255.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
mxa:mango-router max$ yarn wscat --connect wss://api.mngo.cloud/lite-rpc/v1/
yarn run v1.22.15
$ /Users/max/Code/mango-router/node_modules/.bin/wscat --connect wss://api.mngo.cloud/lite-rpc/v1/
Connected (press CTRL+C to quit)
✨  Done in 6.10s.
```

**Comment by mschneider** [comment]
Tried to invoke via regular http rpc request and had no success either, just got a very generic response:
`Welcome to mngo.cloud!`

File: block_priofees/src/block_priofees.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    let jh_priofees_task = tokio::spawn(async move {
+        let sender = priofees_update_sender.clone();
+        'recv_loop: loop {
+            let block = block_stream.recv().await;
+            match block {
+                Ok(block) => {
+                    {
+                        // first do some cleanup
+                        let mut lock = recent_data.write().await;
+                        lock.retain(|slot, _| *slot > slot - SLOTS_TO_RETAIN);
```
Comment (godmodegalactus) [inline block_priofees/src/block_priofees.rs:56]: Do saturating sub, could over flow for local cluster.

File: block_priofees/src/block_priofees.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                    }
+
+                    if !block.commitment_config.is_processed() {
```
Comment (godmodegalactus) [inline block_priofees/src/block_priofees.rs:59]: Could do this test before checking the slot.

File: block_priofees/src/stats_calculation.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        .collect_vec();
+
+    // get stats by CU
+    // e.g. 95 -> 3000
+    let mut dist_fee_by_cu: HashMap<i32, u64> = HashMap::new();
+    let cu_sum: u64 = prio_fees_in_block.iter().map(|x| x.1).sum();
+    let mut agg: u64 = 0;
+    for (prio, cu) in prio_fees_in_block {
+        agg += cu;
+        for p in (0..=100).step_by(5) {
```
Comment (godmodegalactus) [inline block_priofees/src/stats_calculation.rs:41]: This algorithm is suboptimal, it is O(M*N) complex which can be converted to O(M) + O(N) by creating an array of increasing CU after executing the transaction and then in another pass comparing with percentage.

File: lite-rpc/examples/send_udp_too_long.rs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+pub fn main() {
+
+    // dallas, while true; do nc -lu 0.0.0.0 2323; done
+    let server_addr: SocketAddr = "139.178.82.223:2323".parse().unwrap();
+    let socket = UdpSocket::bind("0.0.0.0:0").expect("couldn't bind to address");
+
+    {
+        let buf = [b'x'; 100];
+        println!("sending .. : {}", buf.len());
+        let bytes_sent = socket.send_to(&buf, server_addr).expect("couldn't send data");
```
Comment (godmodegalactus) [inline lite-rpc/examples/send_udp_too_long.rs:13]: Not part of this developement

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

File: block_priofees/src/block_priofees.rs (hunk 4 above)
Comment (grooviegermanikus) [reply block_priofees/src/block_priofees.rs]: good catch!

File: block_priofees/src/stats_calculation.rs (hunk 6 above)
Comment (grooviegermanikus) [reply block_priofees/src/stats_calculation.rs]: updated

**Review by godmodegalactus — APPROVED** [review]

**Pull request merged**